crates/ruff/src/checkers/physical_lines.rs

```diff
@@ -154,9 +156,24 @@ pub fn check_physical_lines(
             }
         }
 
+        // Find the quote delimiter in the line before checking the indentation
+        let (quote_delimiter, occurrences) = find_delimiter(line);
+        if quote_delimiter.is_some() {
+            in_quote = !in_quote;
+        }
```

Comment:
Hmm, I think it's going to be difficult to make this sufficiently performant and robust with the current approach.

One way to think about what's happening here is that we already go through this process of identifying strings when we _tokenize_ the source code, so we're repeating a lot of work here.

Another observation is that we're paying a high cost in the common case to avoid an edge case (most users won't have this rule enabled, most code won't contain tabs, and many tabs won't be in strings).

A few ideas that we could try to leverage here:

1. We could add a `Vec<Range>` to `Indexer` to track all known string locations. Then here, if we detect tab indentation, we could check whether the location is within one of those string ranges. Would need to benchmark to know whether this is sufficiently performant. We already iterate over all tokens there, so grabbing the string ranges is cheap, but we do then have to store them all. (We'd also only need to check "Is this location in a string?" in the event that we detect a tab, so we wouldn't have to pay that price continuously as we do here.)
2. Maybe @MichaReiser has an idea of how this could be achieved within the logical lines framework?

Reply:
That's a good point @charliermarsh, I didn't think about the tokenization process. Where do we do that? I like your Vec<Range> idea. That would be much cleaner.

Reply:
If you take a look at `crates/ruff_python_ast/src/source_code/indexer.rs`, that struct gets a chance to extract meta-information from the token stream, so you could include `Tok::String` in that loop and build up a list of locations. If you look at the caller of _this_ function, it passes in `indexer.commented_lines()` already. So you could change it to take `indexer` directly.

Reply:
Sounds good. Let me try doing that - hopefully the benchmarks improve with that approach. Doing it this way felt off anyway, but that's on me; I didn't even think to check for tokenization. I need to brush up on the whole parsing/lexing/tokenizing process

Reply:
Marking as draft until that's done.